packages/nextjs/src/performance/client.ts

```diff
@@ -149,78 +142,76 @@ export function nextRouterInstrumentation(
-    // (instead of wrapping all of the Router's functions).
-    const routerPrototype = Object.getPrototypeOf(Router.router);
-    fill(routerPrototype, 'changeState', changeStateWrapper);
-  });
-}
+  if (startTransactionOnLocationChange) {
+    Router.events.on('routeChangeStart', (navigationTarget: string) => {
+      const matchedRoute = getNextRouteFromPathname(stripUrlQueryAndFragment(navigationTarget));
 
-type RouterChangeState = (
-  method: string,
-  url: string,
-  as: string,
-  options: Record<string, any>,
-  ...args: any[]
-) => void;
-type WrappedRouterChangeState = RouterChangeState;
+      let transactionName: string;
+      let transactionSource: TransactionSource;
 
-/**
- * Wraps Router.changeState()
- * https://github.com/vercel/next.js/blob/da97a18dafc7799e63aa7985adc95f213c2bf5f3/packages/next/next-server/lib/router/router.ts#L1204
- * Start a navigation transaction every time the router changes state.
- */
-function changeStateWrapper(originalChangeStateWrapper: RouterChangeState): WrappedRouterChangeState {
-  return function wrapper(
-    this: any,
-    method: string,
-    // The parameterized url, ex. posts/[id]/[comment]
-    url: string,
-    // The actual url, ex. posts/85/my-comment
-    as: string,
-    options: Record<string, any>,
-    // At the moment there are no additional arguments (meaning the rest parameter is empty).
-    // This is meant to protect from future additions to Next.js API, especially since this is an
-    // internal API.
-    ...args: any[]
-  ): Promise<boolean> {
-    const newTransactionName = stripUrlQueryAndFragment(url);
-
-    // do not start a transaction if it's from the same page
-    if (startTransaction !== undefined && previousLocation !== as) {
-      previousLocation = as;
-
-      if (activeTransaction) {
-        activeTransaction.finish();
+      if (matchedRoute) {
+        transactionName = matchedRoute;
+        transactionSource = 'route';
+      } else {
+        transactionName = navigationTarget;
+        transactionSource = 'url';
       }
 
       const tags: Record<string, Primitive> = {
         ...DEFAULT_TAGS,
-        method,
-        ...options,
+        from: prevLocationName,
       };
 
-      if (prevTransactionName) {
-        tags.from = prevTransactionName;
+      prevLocationName = transactionName;
+
+      if (activeTransaction) {
+        activeTransaction.finish();
       }
 
-      prevTransactionName = newTransactionName;
-      activeTransaction = startTransaction({
-        name: prevTransactionName,
+      startTransactionCb({
+        name: transactionName,
         op: 'navigation',
         tags,
-        metadata: { source: 'route' },
+        metadata: { source: transactionSource },
       });
-    }
-    return originalChangeStateWrapper.call(this, method, url, as, options, ...args);
-  };
+    });
+  }
+}
+
+function getNextRouteFromPathname(pathname: string): string | void {
```

Comment:
I'm not the biggest fan of union typing with `void` - can we just union type with undefined and have an early return? (see https://github.com/microsoft/TypeScript/issues/42709 on some larger issues with void that I share similar opinions on)

Reply:
Of course! TIL 53b5021eb321553f48ca805f4c1e7c6d0bf11b20